R/external-dependencies.R

```diff
@@ -0,0 +1,111 @@
+external_cli_available <- function(cli) {
+  exit_code <- system(paste("which", cli), ignore.stdout = TRUE, ignore.stderr = TRUE)
+
+  if (exit_code != 0L) {
+    msg <- paste(cli, 'not installed or on $PATH.\n\n')
+    if (cli == "pipx") {
+      msg <- paste0(
+        msg,
+        glue::glue('It can be installed with `install_pipx()`\n\n'),
+        'If already installed, ensure it is on $PATH, e.g. by running',
+        '`pipx ensurepath` or adding `PATH="${PATH}:${HOME}/.local/bin"` to ~/.Renviron'
+      )
+    } else {
+      msg <- paste0(
+        msg,
+        glue::glue('It can be installed with `install_pipx(); install_{cli}()`\n\n'),
+        'If already installed with pipx, ensure it is on $PATH, e.g. by running',
+        '`pipx ensurepath` or adding `PATH="${PATH}:${HOME}/.local/bin"` to ~/.Renviron'
+      )
+    }
+
+    warning(warningCondition(msg, class = "notInstalledWarning"))
+  }
+
+  exit_code == 0L
+}
+
+pipx_available <- function() {
+  external_cli_available(cli = "pipx")
+}
+
+benchconnect_available <- function() {
+  external_cli_available(cli = "benchconnect")
+}
+
+datalogistik_available <- function() {
+  external_cli_available(cli = "datalogistik")
+}
+
+
+#' Install pipx
+#'
+#' Install [pipx](https://pypa.github.io/pipx/), a version of pip that installs
+#' Python packages in isolated environments where they will always be available
+#' regardless of which version of Python is presently on `$PATH`. Especially
+#' useful for installing packages designed to be used via CLIs.
+#'
+#' @export
+install_pipx <- function() {
+  system('pip install pipx && pipx ensurepath', intern = TRUE)
```

Comment:
Ha, this is mostly because I wrote this before I ran into the issues which drove me to `processx::run()` in the first place 😄

We do still have a number of `system()` calls; not just these, but also the very important one [here](https://github.com/voltrondata-labs/arrowbench/blob/main/R/run.R#L434) that runs the script. These are easy to change over, but that one might be trickier—I'm pretty sure there are some subtleties in how that's set up, and I really don't want to bork that one. It's very likely worth changing, but I'd want to make sure Jon has time to do a very thorough review of that particular change.